fuzzers/qemu_systemmode/src/fuzzer.rs

```diff
@@ -0,0 +1,244 @@
+//! A libfuzzer-like fuzzer using qemu for binary-only coverage
+//!
+use core::time::Duration;
+use std::{env, path::PathBuf, process};
+
+use libafl::{
+    bolts::{
+        core_affinity::Cores,
+        current_nanos,
+        launcher::Launcher,
+        rands::StdRand,
+        shmem::{ShMemProvider, StdShMemProvider},
+        tuples::tuple_list,
+        AsSlice,
+    },
+    corpus::{Corpus, InMemoryCorpus, OnDiskCorpus},
+    events::EventConfig,
+    executors::{ExitKind, TimeoutExecutor},
+    feedback_or,
+    feedback_or_fast,
+    feedbacks::{CrashFeedback, MaxMapFeedback, TimeFeedback, TimeoutFeedback},
+    fuzzer::{Fuzzer, StdFuzzer},
+    inputs::{BytesInput, HasTargetBytes},
+    monitors::MultiMonitor,
+    mutators::scheduled::{havoc_mutations, StdScheduledMutator},
+    observers::{TimeObserver, VariableMapObserver},
+    schedulers::{IndexesLenTimeMinimizerScheduler, QueueScheduler},
+    stages::StdMutationalStage,
+    state::{HasCorpus, StdState},
+    Error,
+    //prelude::{SimpleMonitor, SimpleEventManager},
+};
+use libafl_qemu::{
+    //asan::QemuAsanHelper,
+    edges,
+    edges::QemuEdgeCoverageHelper,
+    elf::EasyElf,
+    emu::Emulator,
+    //snapshot::QemuSnapshotHelper,
+    QemuExecutor,
+    QemuHooks,
+    Regs,
+};
+
+fn virt2phys(vaddr: u32, tab: &EasyElf) -> u32 {
```

Comment:
Here you should use the GuestAddr type instead of u32

Reply:
Also about the name, QEMU works with virtual addresses and emu::write_mem takes a virt addr IIRC

Reply:
You bring up a good point.
The function is called that way because it produces the physical load address for a virtual address, as specified in the elf program header.
In my own fork for i used [``cpu_physical_memory_read``/``cpu_physical_memory_write``/``cpu_physical_memory_write_rom``](https://qemu-project.gitlab.io/qemu/devel/loads-stores.html#cpu-physical-memory) instead of [``cpu_memory_rw_debug``](https://qemu-project.gitlab.io/qemu/devel/loads-stores.html#cpu-memory-rw-debug).
Previously i did not use the MMU, and as such did not have to think about the difference.
I think we should offer both options by adding ``emu::read_phys_mem``/``emu::write_phys_mem``.

Reply:
Yes I agree, do you want to add the methods in this PR? Otherwise I'll merge and add them myself

Reply:
This PR. I already implemented them, but need to debug some regressions in the recent commits in the qemu bridge.

Reply:
I was unable to find the root cause of the regression.
The behavior i am seeing is that the VM does not execute instructions after ``vm_start()``.
The usermode fuzzers meanwhile still seem to work.
Using libafl-qemu-bridge [668fc28](https://github.com/AFLplusplus/qemu-libafl-bridge/commit/668fc28b051b78f0b95e051682ddfc884d21fe35) works, but [5bd9736](https://github.com/AFLplusplus/qemu-libafl-bridge/commit/5bd9736af04ea8126532df5fb5e3f82c3fda72fe) does not.
You seem to have merged almost a month of upstream changes and something in there broke it.
I could try to "unroll" your merge into the upstream commits and then bisect it, or you roll back to the working commit for now.
I would suggest the latter, because it might just be a bug that will be fixed anyways.
Unless of course you insist on always being up to date with upstream.

Reply:
I updated again from upstream, please see if it is still broken.
For me is fine to stay at https://github.com/AFLplusplus/qemu-libafl-bridge/commit/668fc28b051b78f0b95e051682ddfc884d21fe35 but just as a temporary workaround.

Maybe the bug is a conflict between the code I added to the single core TCG ops (libafl_cpu_thread_fn et associated) and the new merged code from upstream.

Reply:
It is still broken.
``libafl_cpu_thread_fn`` is not used at all currently, since i just use ``start_vm()``, which launches qemu's normal functions.
I am also not sure how it works.

If you are fine with staying behind for a short time i would suggest merging this PR.
When i get some time during the weekend i can try to find the root cause.
My last commit here reference the working revision, so you could merge right away.

Reply:
It is used to run code in the VCPU thread, as the main thread is for event handling.

The idea was to avoid to stop the VM on resent as it is expensive, while doing everything from the VCPU thread

```rust
use libafl_qemu::*;
use std::env;

extern "C" fn backdoor(pc: GuestAddr, data: u64) {
    let emu = unsafe { (data as *const Emulator).as_ref().unwrap() };
    let cmd: u64 = emu.read_reg(Regs::Rax).unwrap();
    println!("Backdoor invoked from {:#x} with command {}", pc, cmd);
    
    if cmd == 0 {
        println!("Save snapshot");
        emu.save_snapshot("aaa");
    } else if cmd == 1 {
        println!("Break");
        emu.current_cpu().unwrap().trigger_breakpoint();
    }
}

extern "C" fn on_vcpu(mut cpu: CPU) {
    let emu = cpu.emulator();
    
    emu.start(&cpu);

    use std::time::Instant;
    let now = Instant::now();

    for _ in 0..1000 {
        cpu = emu.current_cpu().unwrap(); // ctx switch safe
        emu.start(&cpu);
        
        println!("Load snapshot");
        emu.load_snapshot("aaa");
    }
    
    let elapsed = now.elapsed();
    println!("Elapsed: {:.2?}", elapsed);
    
    //cpu = emu.current_cpu().unwrap();
    //emu.start(&cpu);
    println!("Exiting...");
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let env: Vec<(String, String)> = env::vars().collect();
    let emu = Emulator::new(&args, &env);

    emu.add_backdoor_hook(backdoor, &emu as *const _ as u64);

    emu.set_vcpu_start(on_vcpu);

    unsafe {
        emu.run();
    }
}
```

Reply:
emu::run spawn the VM and the breakpoints, before your patch using qemu_system_debug_request, where simply pausing just the CPU

Reply:
current_cpu was a thing because it is set inside the qemu main loop https://github.com/AFLplusplus/qemu-libafl-bridge/blob/main/accel/tcg/tcg-accel-ops-rr.c#L280

Ofc if you pause the entire VM, there is no current CPU

Reply:
I'm very open to design changes, these things are all really WIP, but we need to clean the code to not have halfway things living alongside other approaches.

Reply:
I just tested it (and my similar changes) and it does not work either.

Reply:
mine apparently runs but i get `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Empty("No entries in corpus", ErrorBacktrace)', src/fuzzer.rs:211:14` like if coverage tracking is not working

Reply:
Ok I think i found the issue

> target/arm/cpu-param.h:# define TARGET_TB_PCREL 1

TARGET_TB_PCREL is a new QEMU option that makes impossible to track precise edge coverage. On x86 is disabled, and I never noticed that arm32 is the only mode with this macro defined to 1

Reply:
yess

```
#ifndef CONFIG_USER_ONLY
# define TARGET_TB_PCREL 1
#endif
```

that's why usermode runs fine

Reply:
e2d34fd5e7a2e7efd3f7d0be8cb320fc62d8334b

This one works

Reply:
Congratulations for finding it so quickly.